pkg/controllers/sandbox/unsandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,267 @@
+			return errors.Join(err, r.updateSandboxStatus(ctx, sbName, succeedScs))
+		}
+		succeedScs = append(succeedScs, sc)
+	}
+	err := r.updateSandboxStatus(ctx, sbName, succeedScs)
+	if err != nil {
+		// when failed to update sandbox status, we will still try to process the sandbox config map
+		return errors.Join(err, r.processConfigMap(ctx))
+	}
+	return r.processConfigMap(ctx)
+}
+
+// processConfigMap will delete the sandbox config map if the sandbox doesn't contain any subclusters,
+// otherwise it will remove unsandbox trigger ID in that config map
+func (r *UnsandboxSubclusterReconciler) processConfigMap(ctx context.Context) error {
+	cmName := r.ConfigMap.Name
+	sb := r.Vdb.GetSandboxStatus(r.ConfigMap.Data[vapi.SandboxNameKey])
+	if sb == nil || len(sb.Subclusters) == 0 {
+		err := r.Client.Delete(ctx, r.ConfigMap)
+		if err != nil {
+			r.Log.Error(err, "failed to delete sandbox config map", "configMapName", cmName)
+			return err
+		}
+		r.Log.Info("Successfully deleted sandbox config map", "configMapName", cmName)
+		return nil
+	}
+	delete(r.ConfigMap.Annotations, vmeta.SandboxControllerUnsandboxTriggerID)
+	err := r.Client.Update(ctx, r.ConfigMap)
+	if err != nil {
+		r.Log.Error(err, "failed to remove unsandbox trigger ID from sandbox config map", "configMapName", cmName)
+		return err
+	}
+	r.Log.Info("Successfully removed unsandbox trigger ID from sandbox config map", "configMapName", cmName)
+
+	return nil
+}
+
+// unsandboxSubcluster will move subclusters from a sandbox to main cluster by calling vclusterOps
+func (r *UnsandboxSubclusterReconciler) unsandboxSubcluster(ctx context.Context, scName string) error {
+	r.SRec.Eventf(r.Vdb, corev1.EventTypeNormal, events.UnsandboxSubclusterStart,
+		"Starting unsandbox subcluster %q", scName)
+	err := r.Dispatcher.UnsandboxSubcluster(ctx,
+		unsandboxsc.WithInitiator(r.InitiatorIP),
+		unsandboxsc.WithSubcluster(scName),
+	)
+	if err != nil {
+		r.SRec.Eventf(r.Vdb, corev1.EventTypeWarning, events.UnsandboxSubclusterFailed,
+			"Failed to unsandbox subcluster %q", scName)
+		return err
+	}
+	r.SRec.Eventf(r.Vdb, corev1.EventTypeNormal, events.UnsandboxSubclusterSucceeded,
+		"Successfully unsandboxed subcluster %q", scName)
+	return nil
+}
+
+// updateSandboxStatus will update sandbox status in vdb
+func (r *UnsandboxSubclusterReconciler) updateSandboxStatus(ctx context.Context, sbName string, scNames []string) error {
+	updateStatus := func(vdbChg *vapi.VerticaDB) error {
+		// update the sandbox's subclusters in sandbox status
+		for i := len(vdbChg.Status.Sandboxes) - 1; i >= 0; i-- {
+			if vdbChg.Status.Sandboxes[i].Name == sbName {
+				vdbChg.Status.Sandboxes[i].Subclusters = vutil.SliceDiff(vdbChg.Status.Sandboxes[i].Subclusters, scNames)
+			}
+			if len(vdbChg.Status.Sandboxes[i].Subclusters) == 0 {
+				vdbChg.Status.Sandboxes = append(vdbChg.Status.Sandboxes[:i], vdbChg.Status.Sandboxes[i+1:]...)
+			}
```

Comment:
We are only updating a single sandbox in this function. Should we have an early out?
```suggestion
			if vdbChg.Status.Sandboxes[i].Name != sbName {
				continue
			}
			vdbChg.Status.Sandboxes[i].Subclusters = vutil.SliceDiff(vdbChg.Status.Sandboxes[i].Subclusters, scNames)
			if len(vdbChg.Status.Sandboxes[i].Subclusters) == 0 {
				vdbChg.Status.Sandboxes = append(vdbChg.Status.Sandboxes[:i], vdbChg.Status.Sandboxes[i+1:]...)
			}
			break
```